Index: x11/nvidia-driver/Makefile
===================================================================
--- x11/nvidia-driver/Makefile	(revision 507040)
+++ x11/nvidia-driver/Makefile	(working copy)
@@ -9,7 +9,7 @@
 # you need to use legacy driver version and install one of corresponding
 # slave ports instead (`x11/nvidia-driver-340' or `x11/nvidia-driver-304').
 
-PORTNAME=	nvidia-driver
+PORTNAME?=	nvidia-driver
 DISTVERSION?=	390.87
 # Always try to set PORTREVISION as it can be overridden by the slave ports
 PORTREVISION?=	3
@@ -19,8 +19,8 @@
 DISTFILES=	${DISTNAME}${EXTRACT_SUFX}
 EXTRACT_ONLY=	${DISTNAME}${EXTRACT_SUFX}
 
-MAINTAINER=	danfe@FreeBSD.org
-COMMENT=	NVidia graphics card binary drivers for hardware OpenGL rendering
+MAINTAINER?=	danfe@FreeBSD.org
+COMMENT?=	NVidia graphics card binary drivers for hardware OpenGL rendering
 
 LICENSE=	NVIDIA
 LICENSE_NAME=	License For Customer Use of NVIDIA Software
@@ -43,7 +43,7 @@
 # since it is still used by legacy drivers.
 MAKE_ENV=	NVIDIA_ROOT=${WRKSRC} X11BASE=${LOCALBASE}
 PLIST_RE=	# empty
-SUB_FILES=	pkg-deinstall pkg-install pkg-message nvidia.conf
+SUB_FILES?=	pkg-deinstall pkg-install pkg-message nvidia.conf
 SUB_PATCHES=	extra-patch-src-Makefile \
 		extra-patch-src-nv-freebsd.h \
 		extra-patch-src-nv-misc.h \
@@ -293,8 +293,30 @@
 		SHLIB_VERSION=${PORTVERSION} \
 		MODULESDIR=${MODULESDIR}
 
+.ifdef LIBGLDIR
+PLIST_SUB+=	LIBGLDIR=${LIBGLDIR}
+.else
+PLIST_SUB+=	LIBGLDIR="@comment "
+.endif
+
+.ifdef NO_LIBMAP
+PLIST_SUB+=	LIBGLMAP="@comment "
+.else
+PLIST_SUB+=	LIBGLMAP=""
+.endif
+
+.ifdef EXTENSIONSLN
+PLIST_SUB+=	EXTENSIONSLN=${EXTENSIONSLN}
+.else
+PLIST_SUB+=	EXTENSIONSLN="@comment "
+.endif
+
 .include <bsd.port.options.mk>
 
+.if ${PORTNAME} != nvidia-headless-driver
+CONFLICTS_INSTALL+=	nvidia-headless-*
+.endif
+
 .if ${PORT_OPTIONS:MLINUX}
 .if ${NVVERSION} >= 390.042 && ${LINUX_DEFAULT} != c6
 USE_LINUX+=	libglvnd
@@ -587,9 +609,33 @@
 		${STAGEDIR}${PREFIX}/lib/libGLESv2-NVIDIA.so.2
 	${RM} -f ${STAGEDIR}${PREFIX}/lib/libGLESv2.so
 .endif
+# nvidia-headless-driver does not want libmap
+.ifndef NO_LIBMAP
 	@${MKDIR} ${STAGEDIR}${PREFIX}/etc/libmap.d/
 	${INSTALL_DATA} ${WRKDIR}/nvidia.conf \
 		${STAGEDIR}${PREFIX}/etc/libmap.d/
+.endif
+# nvidia-headless-driver needs a library path containing unrenamed lib*GL*.so
+.ifdef LIBGLDIR
+	${MKDIR} ${STAGEDIR}${PREFIX}/${LIBGLDIR}
+	${RLN} ${STAGEDIR}${PREFIX}/lib/libGL-NVIDIA.so.1 \
+		${STAGEDIR}${PREFIX}/${LIBGLDIR}/libGL.so.1
+	${RLN} ${STAGEDIR}${PREFIX}/lib/libEGL-NVIDIA.so.1 \
+		${STAGEDIR}${PREFIX}/${LIBGLDIR}/libEGL.so.1
+	${RLN} ${STAGEDIR}${PREFIX}/lib/libGLESv2-NVIDIA.so.2 \
+		${STAGEDIR}${PREFIX}/${LIBGLDIR}/libGLESv2.so.2
+.endif
+# nvidia-headless-driver needs a separate modules dir
+.ifdef EXTENSIONSLN
+	${MKDIR} ${STAGEDIR}${PREFIX}/${EXTENSIONSLN}
+	${RMDIR} ${STAGEDIR}${PREFIX}/${EXTENSIONSLN}
+# ${RLN} (install) has a bug of creating a broken symbolic link if
+# ${EXTENSIONSLN} starts with ${MODULESDIR} as a literal string, regardless of
+# directory separators.
+# EXTENSIONSLN must be chosen carefully by the slave port.
+	${RLN} ${STAGEDIR}${PREFIX}/${MODULESDIR}/extensions/.nvidia \
+		${STAGEDIR}${PREFIX}/${EXTENSIONSLN}
+.endif
 
 post-install-LINUX-on:
 .for f d in ${LINUX_DATA}
Index: x11/nvidia-driver/pkg-plist
===================================================================
--- x11/nvidia-driver/pkg-plist	(revision 507040)
+++ x11/nvidia-driver/pkg-plist	(working copy)
@@ -1,7 +1,10 @@
 bin/nvidia-bug-report.sh
 bin/nvidia-debugdump
 bin/nvidia-smi
-etc/libmap.d/nvidia.conf
+%%LIBGLMAP%%etc/libmap.d/nvidia.conf
+%%LIBGLDIR%%/libGL.so.1
+%%LIBGLDIR%%/libEGL.so.1
+%%LIBGLDIR%%/libGLESv2.so.2
 lib/libGL-NVIDIA.so
 lib/libGL-NVIDIA.so.1
 lib/libEGL-NVIDIA.so
@@ -32,6 +35,7 @@
 %%MODULESDIR%%/drivers/nvidia_drv.so
 %%MODULESDIR%%/extensions/.nvidia/libglx.so
 %%MODULESDIR%%/extensions/.nvidia/libglx.so.1
+%%EXTENSIONSLN%%
 /%%KMODDIR%%/nvidia.ko
 /%%KMODDIR%%/nvidia-modeset.ko
 %%LINUX%%%%LINUXBASE%%/etc/OpenCL/vendors/nvidia.icd
