Index: x11/nvidia-driver/Makefile
===================================================================
--- x11/nvidia-driver/Makefile	(revision 517135)
+++ x11/nvidia-driver/Makefile	(working copy)
@@ -9,7 +9,7 @@
 # you need to use legacy driver version and install one of corresponding
 # slave ports instead (`x11/nvidia-driver-340' or `x11/nvidia-driver-304').
 
-PORTNAME=	nvidia-driver
+PORTNAME?=	nvidia-driver
 DISTVERSION?=	440.31
 # Always try to set PORTREVISION as it can be overridden by the slave ports
 PORTREVISION?=	0
@@ -18,8 +18,8 @@
 DISTNAME=	NVIDIA-FreeBSD-${ARCH_SUFX}-${DISTVERSION}
 DISTFILES=	${DISTNAME}${EXTRACT_SUFX}
 
-MAINTAINER=	danfe@FreeBSD.org
-COMMENT=	NVidia graphics card binary drivers for hardware OpenGL rendering
+MAINTAINER?=	danfe@FreeBSD.org
+COMMENT?=	NVidia graphics card binary drivers for hardware OpenGL rendering
 
 LICENSE_FILE=	${WRKSRC}/doc/license.txt
 .include "${.CURDIR}/../nvidia-driver/Makefile.common"
@@ -58,6 +58,7 @@
 		extra-patch-src-nvidia_pci.c
 DOCSDIR=	${PREFIX}/share/doc/NVIDIA_GLX-1.0
 MODULESDIR=	lib/xorg/modules
+EXTENSIONSDIR?=	${MODULESDIR}/extensions/.nvidia
 PORTDOCS=	*
 
 .if ${NVVERSION} >= 331.013
@@ -93,8 +94,25 @@
 FREEBSD_AGP_DESC=	Use FreeBSD AGP GART driver
 .endif
 
-PLIST_SUB+=	MODULESDIR=${MODULESDIR}
+PLIST_SUB+=	MODULESDIR=${MODULESDIR} \
+		EXTENSIONSDIR=${EXTENSIONSDIR}
 
+.ifdef LIBGLDIR
+PLIST_SUB+=	LIBGLDIR=${LIBGLDIR}
+.else
+PLIST_SUB+=	LIBGLDIR="@comment "
+.endif
+
+.ifdef NO_LIBMAP
+PLIST_SUB+=	LIBGLMAP="@comment "
+.else
+PLIST_SUB+=	LIBGLMAP=""
+.endif
+
+.if ${PORTNAME} != nvidia-headless-driver
+CONFLICTS_INSTALL+=	nvidia-headless-*
+.endif
+
 .include <bsd.port.pre.mk>
 
 pre-patch:
@@ -269,8 +287,30 @@
 		${STAGEDIR}${PREFIX}/lib/libGLESv2-NVIDIA.so.2
 	${RM} ${STAGEDIR}${PREFIX}/lib/libGLESv2.so
 .endif
+# nvidia-headless-driver does not want libmap
+.ifndef NO_LIBMAP
 	@${MKDIR} ${STAGEDIR}${PREFIX}/etc/libmap.d/
 	${INSTALL_DATA} ${WRKDIR}/nvidia.conf \
 		${STAGEDIR}${PREFIX}/etc/libmap.d/
+.endif
+# nvidia-headless-utils needs a library path containing unrenamed lib*GL*.so
+.ifdef LIBGLDIR
+	${MKDIR} ${STAGEDIR}${PREFIX}/${LIBGLDIR}
+	${RLN} ${STAGEDIR}${PREFIX}/lib/libGL-NVIDIA.so.1 \
+		${STAGEDIR}${PREFIX}/${LIBGLDIR}/libGL.so.1
+	${RLN} ${STAGEDIR}${PREFIX}/lib/libEGL-NVIDIA.so.1 \
+		${STAGEDIR}${PREFIX}/${LIBGLDIR}/libEGL.so.1
+	${RLN} ${STAGEDIR}${PREFIX}/lib/libGLESv2-NVIDIA.so.2 \
+		${STAGEDIR}${PREFIX}/${LIBGLDIR}/libGLESv2.so.2
+.endif
+.if ${NVVERSION} < 410.057
+# nvidia-headless-driver needs a separate modules dir
+.if ${EXTENSIONSDIR} != ${MODULESDIR}/extensions/.nvidia
+	${MKDIR} ${STAGEDIR}${PREFIX}/${EXTENSIONSDIR}
+	${MV} ${STAGEDIR}${PREFIX}/${MODULESDIR}/extensions/.nvidia/* \
+		${STAGEDIR}${PREFIX}/${EXTENSIONSDIR}/
+.endif
+.endif
 
+
 .include <bsd.port.post.mk>
Index: x11/nvidia-driver/pkg-plist
===================================================================
--- x11/nvidia-driver/pkg-plist	(revision 517135)
+++ x11/nvidia-driver/pkg-plist	(working copy)
@@ -1,7 +1,10 @@
 bin/nvidia-bug-report.sh
 bin/nvidia-debugdump
 bin/nvidia-smi
-etc/libmap.d/nvidia.conf
+%%LIBGLMAP%%etc/libmap.d/nvidia.conf
+%%LIBGLDIR%%/libGL.so.1
+%%LIBGLDIR%%/libEGL.so.1
+%%LIBGLDIR%%/libGLESv2.so.2
 lib/libEGL-NVIDIA.so
 lib/libEGL-NVIDIA.so.1
 lib/libEGL_nvidia.so
